pkg/mutation/path/parser/node.go

```diff
@@ -32,28 +32,70 @@ type Path struct {
 	Nodes []Node
 }
 
+var _ Node = Path{}
+
 func (r Path) Type() NodeType {
 	return PathNode
 }
 
+func (r Path) DeepCopy() Path {
+	out := Path{
+		Nodes: make([]Node, len(r.Nodes)),
+	}
+	for i := 0; i < len(r.Nodes); i++ {
+		var c Node
+		switch n := r.Nodes[i].(type) {
```

Comment:
We could have the signature for DeepCopy() be:

`DeepCopy() Node`

and add it to the interface, that would eliminate the need for the switch statement.

This is similar to what K8s does for objects, though I think it's a bit different than what we're doing for other parts of mutation.

Curious to hear others' thoughts on returning an interface vs. an object from DeepCopy

Reply:
Yeah I'd be curious what others thoughts are as well as I was trying to decide what to do here. I think if you add `DeepCopy() Node` to the interface, it is not fulfilled by `DeepCopy() Path`, for instance, so one has to actually specify `Node` as the return type.

That means that uses of DeepCopy that want to assign to specific types have to assert/cast (this describes the current use in assignMetadata), but it also means that you don't need type-switch statements if you're copying Nodes without knowing their concrete type.

I went this way after some cursory looking through client-go, where it seems like DeepCopy is defined on all of the concrete types...but it's possible this isn't the way to go or that I read that code a bit wrong :)

Reply:
Ah, you're totally right. K8s uses `DeepCopyObject` when it expects a return type if the `runtime.Object` interface.

Reply:
So by that analog we could add a 

```
func (n TypeGoesHere) DeepCopyNode() Node {
    return n.DeepCopy()
}
```
to avoid the type switch and provide both options?

Reply:
Either way lgtm.

Reply:
I like that, like having a cake and DeepCopying it so you can eat it too

See new patch.